### ISSUE TITLE
Hive Integration

### ISSUE THREAD

**Issue by carver** [ethereum/py-evm#1130]
### What is wrong?

Trinity is not included on hive tests: http://hivetests.ethstats.net/

### How can it be fixed

Information on where to get started:

- See https://github.com/karalabe/hive/blob/master/README.md
- Here's the old pyethereum integration: https://github.com/karalabe/hive/tree/master/clients/pyethereum:develop
- Here's how geth integration looks: https://github.com/karalabe/hive/tree/master/clients/go-ethereum:master

One dockerfile, and one shellscript to init, load blocks, and start rpc services.
The shellscript translates env vars into client-specific configuration

(Info courtesy of @holiman and of course, karalabe)

**Comment by pipermerriam** [comment]
Lets put this in the next milestone.  @carver you want to create that and pick a name?

**Comment by gitcoinbot** [comment]
Issue Status: 1. **Open** 2. Started 3. Submitted 4. Done 

<hr>

__This issue now has a funding of 400.0 DAI (400.0 USD @ $1.0/DAI)  attached to it.__

 * If you would like to work on this issue you can 'start work' [on the Gitcoin Issue Details page](https://gitcoin.co/issue/ethereum/py-evm/1130/1011).
* Want to chip in? Add your own contribution [here](https://gitcoin.co/issue/ethereum/py-evm/1130/1011).
* Questions? Checkout <a href='https://gitcoin.co/help'>Gitcoin Help</a> or the <a href='https://gitcoin.co/slack'>Gitcoin Slack</a>
* $17,432.93 more funded OSS Work available on the [Gitcoin Issue Explorer](https://gitcoin.co/explorer)

**Comment by gitcoinbot** [comment]
Issue Status: 1. **Open** 2. Started 3. Submitted 4. Done 

<hr>

__Workers have applied to start work__.


These users each claimed they can complete the work by 12 months from now.
Please review their action plans below:


1. [cpurta](https://gitcoin.co/profile/cpurta) has applied to start work _(Funders only: [approve worker](https://gitcoin.co/issue/ethereum/py-evm/1130/1011?mutate_worker_action=approve&worker=cpurta) | [reject worker](https://gitcoin.co/issue/ethereum/py-evm/1130/1011?mutate_worker_action=reject&worker=cpurta))_. 
    
    Have already started work on this. Currently have the dockerfile up and running, just need to get the bash script working. Will get a WIP pull request here soon so that the progress can be easily viewed there.

Learn more [on the Gitcoin Issue Details page](https://gitcoin.co/issue/ethereum/py-evm/1130/1011).

**Comment by pipermerriam** [comment]
:+1: @cpurta

**Comment by gitcoinbot** [comment]
@cpurta Hello from Gitcoin Core - are you still working on this issue? Please submit a WIP PR or comment back within the next 3 days or you will be removed from this ticket and it will be returned to an ‘Open’ status. Please let us know if you have questions!
* [x] warning (3 days)
* [ ] escalation to mods (6 days)

Funders only: Snooze warnings for [1 day](https://gitcoin.co/issue/ethereum/py-evm/1130/1011?snooze=1) | [3 days](https://gitcoin.co/issue/ethereum/py-evm/1130/1011?snooze=3) | [5 days](https://gitcoin.co/issue/ethereum/py-evm/1130/1011?snooze=5) | [10 days](https://gitcoin.co/issue/ethereum/py-evm/1130/1011?snooze=10) | [100 days](https://gitcoin.co/issue/ethereum/py-evm/1130/1011?snooze=100)

**Comment by carver** [comment]
> Would it seem necessary to also try and make changes to allow trinity to specify a genesis block?

Yes, that appears to be a requirement for Hive integration. I don't believe anyone else is working on #722 yet.

**Comment by cpurta** [comment]
Also it appears that trinity does not have the option to import blocks similar to geth. Is this something that should also be handled in this issue or perhaps creating a new issue for that feature?

**Comment by carver** [comment]
I don't think it needs its own issue. Should just be a few lines: decode the rlp bytes to an object, and then use the `chain.import_block() method`. (Which is not meant to diminish this issue overall: it's not a tiny project to get Hive integrated)

**Comment by pipermerriam** [comment]
FYI, I added some more explanation to #722 to try and provide some guidance on how it should be implemented.  There is going to be some *figuring out* that needs to be done but hopefully that steers things in the right direction.

**Comment by gitcoinbot** [comment]
@cpurta Hello from Gitcoin Core - are you still working on this issue? Please submit a WIP PR or comment back within the next 3 days or you will be removed from this ticket and it will be returned to an ‘Open’ status. Please let us know if you have questions!
* [x] warning (3 days)
* [ ] escalation to mods (6 days)

Funders only: Snooze warnings for [1 day](https://gitcoin.co/issue/ethereum/py-evm/1130/1011?snooze=1) | [3 days](https://gitcoin.co/issue/ethereum/py-evm/1130/1011?snooze=3) | [5 days](https://gitcoin.co/issue/ethereum/py-evm/1130/1011?snooze=5) | [10 days](https://gitcoin.co/issue/ethereum/py-evm/1130/1011?snooze=10) | [100 days](https://gitcoin.co/issue/ethereum/py-evm/1130/1011?snooze=100)

**Comment by cpurta** [comment]
Just finishing up some of the changes to allow trinity to use a private genesis block. Will be getting a PR up soon. Then will work on importing blocks. Once those are finialized it’ll just be an update to the start script in hive.

**Comment by pipermerriam** [comment]
@cpurta we highly prefer pull requests to be opened as soon as there is anything to show rather than waiting till they are *done*.  This allows us to give preliminary feedback on your approach (which can save you a lot of time in the case your approach isn't viable).  Why don't you open whatever you have now and mark it as a "work in progress".

**Comment by cpurta** [comment]
@carver I am beginning to scope out what is needed for importing blocks to the chain. Would it seems beneficial to add another option to the chain_parser argument (`--import`) to specify a rlp file? Then after the data_dir is initialized to then decode and load the blocks.

**Comment by holiman** [comment]
I'd recommend not mixing import and init. Import should just take blocks from file instead of network

**Comment by holiman** [comment]
(just an opinion)

**Comment by carver** [comment]
Yeah, seems reasonable for it to be its own high level command, like `trinity import /path/to/rlp`

**Comment by cpurta** [comment]
Updated the hive startup script with the correct flags that trinity currently supports and also with the new `genesis` flag. Also roughed out what the import functionality in the script should be like. So once the genesis and import flags are merged the script will probably require just a bit more tweaking.

**Comment by gitcoinbot** [comment]
@cpurta Hello from Gitcoin Core - are you still working on this issue? Please submit a WIP PR or comment back within the next 3 days or you will be removed from this ticket and it will be returned to an ‘Open’ status. Please let us know if you have questions!
* [x] warning (3 days)
* [ ] escalation to mods (6 days)

Funders only: Snooze warnings for [1 day](https://gitcoin.co/issue/ethereum/py-evm/1130/1011?snooze=1) | [3 days](https://gitcoin.co/issue/ethereum/py-evm/1130/1011?snooze=3) | [5 days](https://gitcoin.co/issue/ethereum/py-evm/1130/1011?snooze=5) | [10 days](https://gitcoin.co/issue/ethereum/py-evm/1130/1011?snooze=10) | [100 days](https://gitcoin.co/issue/ethereum/py-evm/1130/1011?snooze=100)

**Comment by cpurta** [comment]
Still working on it. Waiting for the review of the genesis config PR and have started some work on the import portion. Updated the hive startup script to use the correct flags that trinity accepts and using the proposed genesis flag. Once that PR is finalized and merged will just need to get the import chain functionality working then update the startup script.

**Comment by vs77bb** [comment]
Thanks for the update @cpurta - snoozing the bot for 10 days here 🙂

**Comment by pipermerriam** [comment]
@cpurta @vs77bb cpurta did a solid job getting things started on this with the work towards custom genesis files (which I'm going to finish out as it's more nuanced than I originally thought it would be).

After I close that out (hopefully today or tomorrow) we should re-visit this and see what's left to be done for the bounty.

**Comment by vs77bb** [comment]
Sounds good @pipermerriam; anything for @cpurta to do at this point?

**Comment by pipermerriam** [comment]
After working on this for a few days I'm finding it to be much more complex than I initially expected.  After I finish support for custom genesis which I suspect will carry over into next week at minimum I'll re-visit this issue, so `time.sleep(7 days)`

**Comment by gitcoinbot** [comment]
@cpurta Hello from Gitcoin Core - are you still working on this issue? Please submit a WIP PR or comment back within the next 3 days or you will be removed from this ticket and it will be returned to an ‘Open’ status. Please let us know if you have questions!
* [x] warning (3 days)
* [ ] escalation to mods (6 days)

Funders only: Snooze warnings for [1 day](https://gitcoin.co/issue/ethereum/py-evm/1130/1011?snooze=1) | [3 days](https://gitcoin.co/issue/ethereum/py-evm/1130/1011?snooze=3) | [5 days](https://gitcoin.co/issue/ethereum/py-evm/1130/1011?snooze=5) | [10 days](https://gitcoin.co/issue/ethereum/py-evm/1130/1011?snooze=10) | [100 days](https://gitcoin.co/issue/ethereum/py-evm/1130/1011?snooze=100)

**Comment by cpurta** [comment]
@gitcoinbot yes still working on this. Seeing as piper is working on finishing up the the custom genesis support PR there is not much I can do until that is finalized. Once that is done I can update the startup script in hive to use the finalized version of trinity using the custom genesis parameters.

**Comment by pipermerriam** [comment]
I'm hoping to get back into it this week, expecting to close out custom genesis early next week.

**Comment by cpurta** [comment]
Working on the import command for trinity today. I have it mostly stubbed out and working on the rlp decoding and importing that to the chain. Will have a WIP PR up by EOD so it is visible and can make sure that it is needed to try and finish up the Hive integration. Apologies on the delay on this, been a pretty hectic past week.

**Comment by ceresstation** [comment]
Hey @cpurta how is this going?

**Comment by ceresstation** [comment]
Awesome, thanks for the update!

**Comment by pipermerriam** [comment]
@ceresstation #1299 is now merged so you should be unblocked  (thanks for your patience)

**Comment by cpurta** [comment]
I've updated the hive startup script to use the genesis option to start up. Running into an error when starting up with a custom genesis file and working on resolving that issue.

**Comment by carver** [comment]
@cpurta Frank is also trying out Hive functionality, and ran into an issue. Is it the same one? #1527

**Comment by cpurta** [comment]
Yeah same issue that I am seeing.

**Comment by FrankSzendzielarz** [comment]
Re: Hive pls note that the fszendzielarz/hive fork is now defunct and the trinity stuff I added is in the Ethereum/hive repo. I haven't looked at it since the last time I hit a wall with getting the enode, so I think there's still open work in supplying the trinity clients with blockchains to datadir before the containers can be used in consensus tests.
